csi/controller_server.go

```diff
@@ -121,13 +141,30 @@ func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 			return nil, status.Errorf(codes.AlreadyExists, "volume %s size %v differs from requested size %v", existVol.Name, exVolSize, reqVolSizeBytes)
 		}
 
+		if types.VolumeDataSource(existVol.DataSource).IsDataFromAVolume() {
```

Comment:
I am thinking if Longhorn CSI needs to validate clone success or not. If the backup restore is not validated, it shouldn't check the volume clone either.

Reply:
If we don't validate cloning here, Longhorn CSI controller server might return OK for the second CSI `CreateVolume`  calls  even if the clone failed (e.g., the first call timeout waiting for cloning to finish and the second call immediately return OK). As the result, user get an empty volume and Kubernetes never attempt to clone the volume again. If we validate the cloning state here, Kubernetes retry to clone volume with a backoff limit until the volume is cloned. 


IMO, we should implement this logic for backup restore too. How do you think?

Reply:
Is there a retry limit for CSI volume creation? Due to some reason, we choose to wait for the restore success in the attachment stage rather than `CreateVolume` but I don't remember the detail. You may need to try to clone a volume that can spend a long time via CSI first.

Reply:
Trying with 70GB full of data volume

Reply:
I am not sure why we cannot abort a cloning volume but allow deleting a restoring volume. Will the waiting logic inside `VolumeCreate` blocks/fails the CSI volume deletion call?

For the clone retry, YES, we can make it as an enhancement.

Reply:
It is because `VolumeCreate` haven't returned OK together with the volume ID yet so Kubernetes doesn't know which volume to delete. 

In fact, when Kubernetes wants to abort/delete the cloning volume, if it doesn't know volume ID yet, it will make another `VolumeCreate` call to get the volume ID. As the result, even if user manually clean up volume in Longhorn UI, it is recreated:
```
  //    The CO is responsible for cleaning up volumes it provisioned
  //    that it no longer needs. If the CO is uncertain whether a volume
  //    was provisioned or not when a `CreateVolume` call fails, the CO
  //    MAY call `CreateVolume` again, with the same name, to ensure the
  //    volume exists and to retrieve the volume's `volume_id` (unless
  //    otherwise prohibited by "CreateVolume Errors").
```

Reply:
This doesn't happen with restoring volume because as you mentioned: we return OK for restoring volume immediately on the second call from CSI provisioner. E.g., we don't wait for restoring to finish inside `VolumeCreate` call. This is also what we want to do for volume cloning to avoid the above behavior .

Reply:
If the clone wait in `VolumeCreate` will block the deletion, I prefer not to wait for the cloning. "wait" here always means returning error before the clone complete.
For the restore part, I mean we should make the restore behavior consistent with the clone. If we decide to wait for the clone to complete in `VolumeCreate`, we need to add a wait for the restoring as well.

Reply:
Yes, I understand "wait" means return deadline exceeded. In this case, Kubernetes will not have volume ID thus cannot delete. 

Yes, I agree that we shouldn't wait inside VolumeCreate for both clone and backup restore. I already updated the PR to reflect this.

Reply:
Can you add a comment explaining why Longhorn won't wait for clone/restore complete? Then this PR is good to go.

Reply:
Yes, I added the comment. Thank you